src/ffi.rs

```diff
@@ -578,6 +578,180 @@ pub unsafe extern "C" fn context_get_result(
+/// - `errbuf_len`: a pointer to the length of the error message buffer.
+///
+/// # Returns
+///
+/// Returns an integer value indicating the validation result:
+/// - ATC_ROUTER_EXPRESSION_VALIDATE_OK(0) if validation is passed.
+/// - ATC_ROUTER_EXPRESSION_VALIDATE_FAILED(1) if validation is failed.
+/// - ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL(2) if the provided fields buffer is not enough.
+///
+/// If `fields_buf` is null and `fields_len` or `fields_total` is non-null, it will write
+/// the required buffer length and the total number of fields to the provided pointers.
+/// If `fields_buf` is non-null, and `fields_len` is enough for the required buffer length,
+/// it will write the used fields to the buffer separated by '\0' and the total number of fields
+/// to the `fields_total`, and `fields_len` will be updated with the total buffer length.
+/// If `fields_buf` is non-null, and `fields_len` is not enough for the required buffer length,
+/// it will write the required buffer length to the `fields_len`, and the total number of fields
+/// to the `fields_total`, and return `2`.
+/// If `operators` is non-null, it will write the used operators with bitflags to the provided pointer.
+/// The bitflags is defined by `BinaryOperatorFlags` and it must not contain any bits from `BinaryOperatorFlags::UNUSED`.
+///
+///
+/// # Panics
+///
+/// This function will panic when:
+///
+/// - `atc` doesn't point to a valid C-style string.
+/// - `fields_len` and `fields_total` are null when `fields_buf` is non-null.
+///
+/// # Safety
+///
+/// Violating any of the following constraints will result in undefined behavior:
+///
+/// - `atc` must be a valid pointer to a C-style string, must be properly aligned,
+///    and must not have '\0' in the middle.
+/// - `schema` must be a valid pointer returned by [`schema_new`].
+/// - `fields_buf` must be a valid to write for `fields_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `fields_len` must be a valid to write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `fields_total` must be a valid to write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `operators` must be a valid to write for `size_of::<u64>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `errbuf` must be valid to read and write for `errbuf_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned.
+/// - `errbuf_len` must be valid to read and write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned.
+/// - If `fields_buf` is non-null, `fields_len` and `fields_total` must be non-null.
+/// - If `fields_buf` is null, `fields_len` and `fields_total` can be non-null
+///   for writing required buffer length and total number of fields.
+#[cfg(feature = "expr_validation")]
+#[no_mangle]
+pub unsafe extern "C" fn expression_validate(
+    atc: *const u8,
```

Comment:
any reason to use `u8` for strings here, and `i8` elsewhere?

Reply:
According to the [document](https://doc.rust-lang.org/std/ffi/type.c_char.html), `ffi::c_char` is defined as `i8`. However, the document also mentioned that On modern architectures this type will always be either `i8` or `u8`.

Choosing `u8` here is just to keep consistency with Rust internal bytes represent like [`String::as_bytes`](https://doc.rust-lang.org/std/string/struct.String.html#method.as_bytes). 

P.S. I saw all the FFI interfaces in this project have no consistent way to represent `c_char`. E.g. `fields` is `u8` in `router_get_fields` and `i8` in `context_add_value`.

Reply:
@Oyami-Srk In general, we tries to use `u8` for `char` whenever possible, except when dealing with char provided by LuaJIT.

This is because LuaJIT converts string to `const char *`, and this is a signed value. If we accept `u8` LuaJIT will complain when attempting to convert a string.